File: src/Board.java
import java.util.ArrayList;

/**
 * @author	Andrew Sytsma <asytsma@purdue.edu>
 */
public class Board
{
	/**
	 * Holds the value of the in progress state.
	 */
	private final int IN_PROGRESS = 0;
	
	/**
	 * Holds the value of the game over state.
	 */
	private final int GAME_OVER = 1;
	
	/**
	 * Holds the value of the tie state.
	 */
	private final int TIE = 2;
	
	/**
	 * Holds the current player's turn.
	 */
	private int turn;
	
	/**
	 * Holds a reference to the board.
	 */
	private Piece[][][] board;
	
	/**
	 * Holds a reference to the first player.
	 */
	private Player player1;
	
	/**
	 * Holds a reference to the second player.
	 */
	private Player player2;
	
	/**
	 * Holds all the available moves left on the board still
	 */
	private ArrayList<Location> availableMoves;
	
	/**
	 * Holds the current state of the game.
	 */
	private int state;
	
	/**
	 * Board constructor initializes fields
	 * 
	 * @param player1	Holds a reference to the first player.
	 * @param player2	Holds a reference to the second player.
	 */
	public Board(Player player1, Player player2)
	{
		this.turn = 1;
		this.board = new Piece[4][4][4];
		this.availableMoves = new ArrayList<>();
		
		//populate the available moves
		for (int x = 0; x < 4; x++)
		{
			for (int y = 0; y < 4; y++)
			{
				for (int z = 0; z < 4; z++)
				{
					availableMoves.add(new Location(x, y, z));
				}	//end for
			}	//end for
		}	//end for
		
		this.player1 = player1;
		this.player2 = player2;
		this.state = this.IN_PROGRESS;
	}	//end of Board constructor
	
	/**
	 * Returns whether or not the location is taken.
	 * True  = Taken
	 * False = Not Taken
	 * 
	 * @param location	Holds the location to be checked.
	 * @return			Returns whether or not the location is taken.
	 */
	private boolean isTaken(Location location)
	{
		//check if the location is taken
		if (this.board[location.getX()][location.getY()][location.getZ()] == null)	//the location is not taken
		{
			return false;
		}	//end if
		
		return true;
	}	//end of isValid method
	
	/**
	 * Returns a reference to the current player.
	 * 
	 * @return	Returns a reference to the current player.
	 */
	public Player getCurrentPlayer()
	{
		//check what player's turn it is
		if (this.turn == 1)	//player1's turn
		{
			return this.player1;
		}
		else	//player2's turn
		{
			return this.player2;
		}	//end if
	}	//end of getCurrentPlayer method
	
	/**
	 * Returns the turn number.
	 * 1 = Player1
	 * 2 = Player2
	 * 
	 * @return	Returns the turn number.
	 */
	public int getTurn()
	{
		return this.turn;
	}	//end getTurn method
	
	/**
	 * Makes a player's move.
	 * 
	 * @param location	Holds the location of the move.
	 */
	public void makeMove(Location location)
	{
		//check if the game is over
		if (this.state != this.IN_PROGRESS)	//the game is over
		{
			return;
		}	//end if
		
		//check if the location is taken
		if (isTaken(location))
		{
			return;
		}	//end if
		
		Piece piece;	//holds the piece to be played
		
		//check what player's turn it is
		if (this.turn == 1)	//player1's turn
		{
			piece = new X(location);
		}
		else	//player2's turn
		{
			piece = new O(location);
		}	//end if
		
		//put the piece on the board
		this.board[location.getX()][location.getY()][location.getZ()] = piece;
		
		//remove the piece from the list
		this.availableMoves.remove(location);
		
		int state = isGameOver(piece.getClass());	//holds the state of the game
		
		//check if the game is over
		if (state == this.IN_PROGRESS)	//game is not over
		{
			//swap the turns
			if (this.turn == 1)	//it was player1's turn
			{
				this.turn = 2;
			}
			else	//it was player2's turn
			{
				this.turn = 1;
			}	//end if
		}
	}	//end of makeMove method
	
	/**
	 * Returns a number representing the state of the game.
	 * IN_PROGRESS = Not Game Over
	 * GAME_OVER = Game Over
	 * TIE = Tie
	 * 
	 * @return	Returns a number representing the state of the game.
	 */
	public int getState()
	{
		return this.state;
	}	//end of getState method
	
	/**
	 * Returns a number representing the state of the game.
	 * IN_PROGRESS = Not Game Over
	 * GAME_OVER = Game Over
	 * TIE = Tie
	 * 
	 * @return	Returns a number representing the state of the game.
	 */
	private int isGameOver(Class piece)
	{
		//check the 2D rows
		if (check2DRows(piece))	//a win was found
		{
			return this.GAME_OVER;
		}	//end if
		
		//check the 2D columns
		if (check2DColumns(piece))	//a win was found
		{
			return this.GAME_OVER;
		}	//end if
		
		//check the 2D diagonals
		if (check2DDiagonals(piece))	//a win was found
		{
			return this.GAME_OVER;
		}	//end if
		
		//check the 3D rows
		if (check3DRows(piece))	//a win was found
		{
			return this.GAME_OVER;
		}	//end if
		
		//check the 3D diagonals
		if (check3DDiagonals(piece))	//a win was found
		{
			return this.GAME_OVER;
		}	//end if
		
		//check for a tie
		if (isTie())	//a tie was found
		{
			return this.TIE;
		}	//end if
	
		return this.IN_PROGRESS;
	}	//end of isGameOver method
	
	/**
	 * Returns whether or not there is a tie.
	 * True  = Tie
	 * False = No Tie 
	 * 
	 * @return	Returns whether or not there is a tie.
	 */
	private boolean isTie()
	{
		//loop through the board to see if all the spaces are filled
		for (int x = 0; x < 4; x += 1)
		{
			for (int y = 0; y < 4; y += 1)
			{
				for (int z = 0; z < 4; z += 1)
				{
					//check if the space is empty
					if (this.board[x][y][z] == null)	//the space is empty
					{
						return false;
					}	//end if
				}	//end for
			}	//end for
		}	//end for
		
		return true;
	}	//end isTie method
	
	/**
	 * Returns whether or not there is a win in the 2D rows.
	 * True  = Win
	 * False = No Win
	 * 
	 * @return	Returns whether or not there is a win in the 2D rows.
	 */
	private boolean check2DRows(Class piece)
	{
		//loop through the 2D rows
		for (int z = 0; z < 4; z += 1)
		{
			for (int y = 0; y < 4; y += 1)
			{
				boolean found = true;	//holds if a win was found
				
				for (int x = 0; x < 4; x += 1)
				{
					//check the piece
					if (!piece.isInstance(this.board[x][y][z]))	//bad piece
					{
						found = false;
						break;
					}	//end if
				}	//end for
				
				//check if a win was found
				if (found)	//a win was found
				{
					return true;
				}	//end if
			}	//end for
		}	//end for
		
		return false;
	}	//end of checkRows method
	
	/**
	 * Returns whether or not there is a win in the 2D columns.
	 * True  = Win
	 * False = No Win
	 * 
	 * @return	Returns whether or not there is a win in the 2D columns.
	 */
	private boolean check2DColumns(Class piece)
	{
		//loop through the 2D columns
		for (int z = 0; z < 4; z += 1)
		{
			for (int x = 0; x < 4; x += 1)
			{
				boolean found = true;	//holds if a win was found
				
				for (int y = 0; y < 4; y += 1)
				{
					//check the piece
					if (!piece.isInstance(this.board[x][y][z]))	//bad piece
					{
						found = false;
						break;
					}	//end if
				}	//end for
				
				//check if a win was found
				if (found)	//a win was found
				{
					return true;
				}
			}	//end for
		}	//end for
		
		return false;
	}	//end of check2DColumns method
	
	/**
	 * Returns whether or not there is a win in the 2D diagonals.
	 * True  = Win
	 * False = No Win
	 * 
	 * @return	Returns whether or not there is a win in the 2D diagonals.
	 */
	private boolean check2DDiagonals(Class piece)
	{
		//loop through the 2D diagonals
		for (int z = 0; z < 4; z += 1)
		{
			boolean found = true;	//holds if a win was found
			
			//check the 2D diagonals
			if (piece.isInstance(this.board[0][0][z]))	//left diagonal
			{
				for (int num = 1; num < 4; num += 1)
				{
					//check a piece
					if (!piece.isInstance(this.board[num][num][z]))	//bad piece
					{
						found = false;
						break;
					}	//end if
				}	//end for
			}
			else if (piece.isInstance(this.board[3][0][z]))	//right diagonal
			{
				for (int num = 1; num < 4; num += 1)
				{
					//check a piece
					if (!piece.isInstance(this.board[3 - num][num][z]))	//bad piece
					{
						found = false;
						break;
					}	//end if
				}	//end for
			}
			else	//no diagonal
			{
				found = false;
			}	//end if
			
			//check if a win was found
			if (found)	//a win was found
			{
				return true;
			}
		}	//end for
		
		return false;
	}	//end of check2DDiagonals method
	
	/**
	 * Returns whether or not there is a win in the 3D rows.
	 * True  = Win
	 * False = No Win
	 * 
	 * @return	Returns whether or not there is a win in the 3D rows.
	 */
	private boolean check3DRows(Class piece)
	{
		//loop through the 3D rows
		for (int x = 0; x < 4; x += 1)
		{
			for (int y = 0; y < 4; y += 1)
			{
				boolean found = true;	//holds if a win was found
				
				for (int z = 0; z < 4; z += 1)
				{
					//check a piece
					if (!piece.isInstance(this.board[x][y][z]))	//bad piece
					{
						found = false;
						break;
					}	//end if
				}	//end for
				
				//check if a win was found
				if (found)	//a win was found
				{
					return true;
				}	//end if
			}	//end for
		}	//end for
		
		return false;
	}	//end of check3DRows method
	
	/**
	 * Returns whether or not there is a win in the 3D diagonals.
	 * True  = Win
	 * False = No Win
	 * 
	 * @return	Returns whether or not there is a win in the 3D diagonals.
	 */
	private boolean check3DDiagonals(Class piece)
	{
		//check the 3D diagonals
		if (piece.isInstance(this.board[0][0][0]))	//top-left diagonal
		{
			for (int num = 1; num < 4; num += 1)
			{
				//check a piece
				if (!piece.isInstance(this.board[num][num][num]))	//bad piece
				{
					return false;
				}	//end if
			}	//end for
		}
		else if (piece.isInstance(this.board[3][0][0]))	//top-right diagonal
		{
			for (int num = 1; num < 4; num += 1)
			{
				//check a piece
				if (!piece.isInstance(this.board[3 - num][num][num]))	//bad piece
				{
					return false;
				}	//end if
			}	//end for
		}
		else if (piece.isInstance(this.board[0][3][0]))	//bottom-left diagonal
		{
			for (int num = 1; num < 4; num += 1)
			{
				//check a piece
				if (!piece.isInstance(this.board[num][3 - num][num]))	//bad piece
				{
					return false;
				}	//end if
			}	//end for
		}
		else if (piece.isInstance(this.board[3][3][0]))	//bottom-right diagonal
		{
			for (int num = 1; num < 4; num += 1)
			{
				//check a piece
				if (!piece.isInstance(this.board[3 - num][3 - num][num]))	//bad piece
				{
					return false;
				}	//end if
			}	//end for
		}
		else	//no diagonal
		{
			return false;
		}	//end if
		
		return true;
	}	//end of check3DDiagonals method
	
	/**
	 * Get the available moves still left on the board.
	 * 
	 * @return	The available moves
	 */
	public ArrayList<Location> getAvailableMoves()
	{
		return this.availableMoves;
	}	//end of getAvailableMoves method
	
	/**
	 * Returns a reference to the board array.
	 * 
	 * @return	Returns a reference to the board array.
	 */
	public Piece[][][] getBoard()
	{
		return this.board;
	}	//end of getBoard method
	
	/**
	 * Returns a number representing the state of the game.
	 * IN_PROGRESS = Not Game Over
	 * GAME_OVER = Game Over
	 * TIE = Tie
	 * 
	 * @param testBoard	The Board to test.
	 * @param turn		A number representing the player's turn.
	 * @return			Returns a number representing the state of the game.
	 */
	public int testBoard(Piece[][][] testBoard, int turn)
	{
		Piece[][][] oldBoard = this.board;	//Holds a reference to the old board.
		this.board = testBoard;				//Sets the board to the test board.
		int result;							//Holds the result of the test.
		
		//check which player's turn it is
		if (turn == 1)	//player1's turn
		{
			//check for a win with X's
			result = isGameOver(new X(new Location(0, 0, 0)).getClass());
		}
		else	//player2's turn
		{
			//check for a win with O's
			result = isGameOver(new O(new Location(0, 0, 0)).getClass());
		}	//end if
		
		this.board = oldBoard;	//reset the board
		
		return result;
	}	//end of testBoard method
}	//end of Board class

File: WebContent/classes/Board.php
<?php
	require_once($homeDir . "classes/Location.php");
	
	class Board
	{
		private $IN_PROGRESS = 0;
		private $GAME_OVER = 1;
		private $TIE = 2;
		private $X_PIECE = 1;
		private $O_PIECE = 2;
		private $EMPTY_SPACE = 0;
		private $turn;
		private $board;
		private $player1;
		private $player2;
		private $availableMoves;
		private $state;
		
		
		public function __construct(Player $p1, Player $p2)
		{
			$this->turn = 1;
			$this->board = array
			(
				array
				(
					array(0, 0, 0, 0),
					array(0, 0, 0, 0),
					array(0, 0, 0, 0),
					array(0, 0, 0, 0)
				),
				array
				(
					array(0, 0, 0, 0),
					array(0, 0, 0, 0),
					array(0, 0, 0, 0),
					array(0, 0, 0, 0)
				),
				array
				(
					array(0, 0, 0, 0),
					array(0, 0, 0, 0),
					array(0, 0, 0, 0),
					array(0, 0, 0, 0)
				),
				array
				(
					array(0, 0, 0, 0),
					array(0, 0, 0, 0),
					array(0, 0, 0, 0),
					array(0, 0, 0, 0)
				)
			);
			$this->player1 = $p1;
			$this->player2 = $p2;
			$this->availableMoves = array();
			
			for ($x = 0; $x < 4; $x++)
			{
				for ($y = 0; $y < 4; $y++)
				{
					for ($z = 0; $z < 4; $z++)
					{
						$this->availableMoves[] = array($x, $y, $z);
					}	//end for
				}	//end for
			}	//end for
			
			$this->state = $this->IN_PROGRESS;
		}	//end of Board constructor
		
		private function isTaken($xLoc, $yLoc, $zLoc)
		{			
			if ($this->board[$xLoc][$yLoc][$zLoc] === $this->EMPTY_SPACE)
			{
				return false;
			}	//end if
			
			return true;
		}	//end of isTaken method
		
		public function getCurrentPlayer()
		{
			if ($this->turn === 1)
			{
				return $this->player1;
			}	//end if
			
			return $this->player2;
		}	//end of getCurrentPlayer method
		
		public function getTurn()
		{
			return $this->turn;
		}	//end of getTurn method
		
		public function makeMove($coordinate)
		{
			$xLoc = $coordinate[0];
			$yLoc = $coordinate[1];
			$zLoc = $coordinate[2];
			if ($this->state != $this->IN_PROGRESS)
			{
				return;
			}	//end if
			
			if ($this->isTaken($xLoc, $yLoc, $zLoc))
			{
				return;
			}	//end if
			
			$piece = $this->O_PIECE;
			
			if ($this->turn === 1)
			{
				$piece = $this->X_PIECE;
			}	//end if
			
			$this->board[$xLoc][$yLoc][$zLoc] = $piece;
			
			for($x = 0; $x < count($this->availableMoves); $x++)
			{
				$c = $this->availableMoves[$x];
				if($c[0] === $xLoc && $c[1] === $yLoc && $c[2] === $zLoc)
				{
					unset($this->availableMoves[$x]);
					$this->availableMoves = array_values($this->availableMoves);
					break;
				}
			}
			
			$result = $this->isGameOver($piece);
			
			if ($result === $this->IN_PROGRESS)
			{
				if ($this->turn === 1)
				{
					$this->turn = 2;
				}
				else
				{
					$this->turn = 1;
				}	//end if
			}
			else
			{
				$this->state = $result;
			}	//end if
		}	//end of makeMove method
		
		public function getState()
		{
			return $this->state;
		}	//end of getState method
		
		private function isGameOver($piece)
		{
			/*if ($this->check2DRows($piece))
			{
				return $this->GAME_OVER;
			}	//end if
			
			if ($this->check2DColumns($piece))
			{
				return $this->GAME_OVER;
			}	//end if
			
			if ($this->check2DDiagonals($piece))
			{
				return $this->GAME_OVER;
			}	//end if
			
			if ($this->check3DRows($piece))
			{
				return $this->GAME_OVER;
			}	//end if
			
			if ($this->check3DDiagonals($piece))
			{
				return $this->GAME_OVER;
			}	//end if */
			
			if($this->checkBoard($piece))
			{
				return $this->GAME_OVER;
			}
			
			if ($this->isTie())
			{
				return $this->TIE;
			}	//end if
			
			return $this->IN_PROGRESS;
		}	//end of isGameOver method
		
		private function isTie()
		{
			
			/*for ($x = 0; $x < 4; $x++)
			{
				for ($y = 0; $y < 4; $y++)
				{
					for ($z = 0; $z < 4; $z++)
					{
						if ($this->board[$x][$y][$z] === $this->EMPTY_SPACE)
						{
							return false;
						}	//end if
					}	//end for
				}	//end for
			}	//end for*/
			
			
			if(count($this->availableMoves) === 0)
			{
				return true;
			}
			return false;
			
			//return true;
		}	//end of isTie method
		
		private function check2DBoard($board, $piece)
		{
			$found = true;
			
			//search all columns
			for($x = 0; $x < 4; $x++)
			{
				$found = true;
				
				//search all rows in column
				for($y = 0; $y < 4; $y++)
				{
					//not continuous
					if($board[$x][$y] !== $piece)
					{
						$found = false;
						break;
					}
				}
				
				//we found one
				if($found)
				{
					return true;
				}
			}
			
			//search all rows
			for($y = 0; $y < 4; $y++)
			{
				$found = true;
				
				//search all columns in row
				for($x = 0; $x < 4; $x++)
				{
					//not continuous
					if($board[$x][$y] !== $piece)
					{
						$found = false;
						break;
					}
				}
				
				//we found one
				if($found)
				{
					return true;
				}
			}
			
			//check top left to bottom right
			$found = true;
			for($x = 0; $x < 4; $x++)
			{
				//not continuous
				if($board[$x][$x] !== $piece)
				{
					$found = false;
					break;
				}
			}
			
			//we found one
			if($found)
			{
				return true;
			}
			
			//check bottom left to top right
			$found = true;
			for($x = 0; $x < 4; $x++)
			{
				//not continuous
				if($board[$x][3 - $x] !== $piece)
				{
					$found = false;
					break;
				}
			}
			
			return $found;
		}
		
		private function checkBoard($piece)
		{
			$board2D = array
			(
				array(0, 0, 0, 0),
				array(0, 0, 0, 0),
				array(0, 0, 0, 0),
				array(0, 0, 0, 0)
			);
			
			//search 2D top to bottom
			for($z = 0; $z < 4; $z++)
			{
				//get 2D board
				for($y = 0; $y < 4; $y++)
				{
					for($x = 0; $x < 4; $x++)
					{
						$board2D[$x][$y] = $this->board[$x][$y][$z];
					}
				}
				
				//this board has a connection
				if($this->check2DBoard($board2D, $piece))
				{
					return true;
				}
			}
			
			//search 2D left to right
			for($x = 0; $x < 4; $x++)
			{
				//get 2D board
				for($z = 0; $z < 4; $z++)
				{
					for($y = 0; $y < 4; $y++)
					{
						$board2D[$y][$z] = $this->board[$x][$y][$z];
					}
				}
				
				//this board has a connection
				if($this->check2DBoard($board2D, $piece))
				{
					return true;
				}
			}
			
			//search 2D back to front
			for($y = 0; $y < 4; $y++)
			{
				//get 2D board
				for($z = 0; $z < 4; $z++)
				{
					for($x = 0; $x < 4; $x++)
					{
						$board2D[$x][$z] = $this->board[$x][$y][$z];
					}
				}
				
				//this board has a connection
				if($this->check2DBoard($board2D, $piece))
				{
					return true;
				}
			}
			
			//4 diagonals
			//top left back to bottom right front
			$found = true;
			for($x = 0; $x < 4; $x++)
			{
				//not continuous
				if($this->board[$x][$x][$x] !== $piece)
				{
					$found = false;
					break;
				}
			}
			
			//we found one
			if($found)
			{
				return true;
			}
			
			//bottom left back to top right front
			$found = true;
			for($x = 0; $x < 4; $x++)
			{
				//not continuous
				if($this->board[$x][$x][3 - $x] !== $piece)
				{
					$found = false;
					break;
				}
			}
			
			//we found one
			if($found)
			{
				return true;
			}
			
			//top left front to bottom right back
			$found = true;
			for($x = 0; $x < 4; $x++)
			{
				//not continuous
				if($this->board[$x][3 - $x][$x] !== $piece)
				{
					$found = false;
					break;
				}
			}
			
			//we found one
			if($found)
			{
				return true;
			}
			
			//bottom left front to top right back
			$found = true;
			for($x = 0; $x < 4; $x++)
			{
				//not continuous
				if($this->board[$x][3 - $x][3 - $x] !== $piece)
				{
					$found = false;
					break;
				}
			}
			
			return $found;
		}
		
		private function check2DRows($piece)
		{
			for ($z = 0; $z < 4; $z++)
			{
				for ($y = 0; $y < 4; $y++)
				{
					$found = true;
					
					for ($x = 0; $x < 4; $x++)
					{
						if ($piece !== $this->board[$x][$y][$z])
						{
							$found = false;
							break;
						}	//end if
					}	//end for
					
					if ($found)
					{
						return true;
					}	//end if
				}	//end for
			}	//end for
			
			return false;
		}	//end of checkRows method
		
		private function check2DColumns($piece)
		{
			for ($z = 0; $z < 4; $z++)
			{
				for ($x = 0; $x < 4; $x++)
				{
					$found = true;
					
					for ($y = 0; $y < 4; $y++)
					{
						if ($piece !== $this->board[$x][$y][$z])
						{
							$found = false;
							break;
						}	//end if
					}	//end for
					
					if ($found)
					{
						return true;
					}	//end if
				}	//end for
			}	//end for
			
			return false;
		}	//end of check2DColumns method
		
		private function check2DDiagonals($piece)
		{
			for ($z = 0; $z < 4; $z++)
			{
				$found = true;
				
				if ($piece === $this->board[0][0][$z])
				{
					for ($num = 1; $num < 4; $num++)
					{
						if ($piece !== $this->board[$num][$num][$z])
						{
							$found = false;
							break;
						}	//end if
					}	//end for
				}
				else if ($piece === $this->board[3][0][$z])
				{
					for ($num = 1; $num < 4; $num++)
					{
						if ($piece !== $this->board[3 - $num][$num][$z])
						{
							$found = false;
							break;
						}	//end if
					}	//end for
				}
				else
				{
					$found = false;
				}	//end if
				
				if ($found)
				{
					return true;
				}
			}	//end for
			
			return false;
		}	//end of check2DDiagonals method
		
		private function check3DRows($piece)
		{
			for ($x = 0; $x < 4; $x++)
			{
				for ($y = 0; $y < 4; $y++)
				{
					$found = true;
					
					for ($z = 0; $z < 4; $z++)
					{
						if ($piece !== $this->board[$x][$y][$z])
						{
							$found = false;
							break;
						}	//end if
					}	//end for
					
					if ($found)
					{
						return true;
					}	//end if
				}	//end for
			}	//end for
			
			return false;
		}	//end of check3DRows method
		
		private function check3DDiagonals($piece)
		{
			if ($piece === $this->board[0][0][0])
			{
				for ($num = 1; $num < 4; $num++)
				{
					if ($piece !== $this->board[$num][$num][$num])
					{
						return false;
					}	//end if
				}	//end for
			}
			else if ($piece === $this->board[3][0][0])
			{
				for ($num = 1; $num < 4; $num++)
				{
					if ($piece !== $this->board[3 - $num][$num][$num])
					{
						return false;
					}	//end if
				}	//end for
			}
			else if ($piece === $this->board[0][3][0])
			{
				for ($num = 1; $num < 4; $num++)
				{
					if ($piece !== $this->board[$num][3 - $num][$num])
					{
						return false;
					}	//end if
				}	//end for
			}
			else if ($piece === $this->board[3][3][0])
			{
				for ($num = 1; $num < 4; $num++)
				{
					if ($piece !== $this->board[3 - $num][3 - $num][$num])
					{
						return false;
					}	//end if
				}	//end for
			}
			else
			{
				return false;
			}	//end if
			
			return true;
		}	//end of check3DDiagonals method
		
		public function getAvailableMoves()
		{
			return $this->availableMoves;
		}	//end of getAvailableMoves method
		
		public function getBoard()
		{
			return $this->board;
		}	//end of getBoard method
		
		public function testBoard($testBoard, $t)
		{
			$oldBoard = $this->board;
			$this->board = $testBoard;
			$result;
			
			if ($t == 1)
			{
				$result = $this->isGameOver($this->X_PIECE);
			}
			else
			{
				$result = $this->isGameOver($this->O_PIECE);
			}	//end if
			
			$this->board = $oldBoard;
			
			return $result;
		}	//end of testBoard method
	}	//end of Board class
?>